### PR TITLE
Church Commands Version 1

### PR DESCRIPTION
**Pull request by yuqi-zhang**
Add initial Church of VileBot functionality, including:
- donate (!donate)
- viewing total donations (!churchtotal)
- see top 4 donors and titles (!topdonors)
- set your title if you're one of the top 4 donors (!settitle <noun>)
- have VileBot speak in Zalgo instead of normal (!gospel)

### CODE REVIEW

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        Double karma;
+        try
+        {
+            karma = jedis.zscore( keyOfChurchDonorSortedSet, noun );
+        }
+        finally
+        {
+            pool.returnResource( jedis );
+        }
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        if ( karma == null )
+        {
+            return null;
+        }
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        Long rank;
+        try
+        {
+            rank = jedis.zrevrank( keyOfChurchDonorSortedSet, noun );
+        }
+        finally
+        {
+            pool.returnResource( jedis );
+        }
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        if ( rank == null )
+        {
+            return null;
+        }
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+     * Get nouns from karma ranks.
+     *
+     * @param lower The lower rank to get the nouns of.
+     * @param upper The upper rank to get the nouns of.
+     * @return String The noun iff the rank exists, else null.
+     */
+    public static Set<String> getDonorsByRanks( long lower, long upper )
+    {
+        Set<String> nouns;
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            return null;
+        }
+
+        return nouns;
+    }
+
+    public static boolean removeDonor( String noun )
+    {
+        Long existed;
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        Jedis jedis = pool.getResource();
+        try
+        {
+            existed = jedis.zrem( keyOfChurchDonorSortedSet, noun );
+        }
+        finally
+        {
+            pool.returnResource( jedis );
+        }
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: remove new line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        }
+        return true;
+    }
+
+    public static long getTotalDonations() {
+        Jedis jedis = pool.getResource();
+        long totalKarma;
+        try
+        {
+            Set<String> members = jedis.zrange(keyOfChurchDonorSortedSet, 0, -1);
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            totalKarma = sum(members, jedis);
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        }
+        return totalKarma;
+    }
+
+    public static long getTotalNonDonations() {
+         Jedis jedis = pool.getResource();
+         long totalKarma;
+         try
+         {
+             Set<String> members = jedis.zrange(keyOfChurchSortedSet, 0, -1);
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+             totalKarma = sum(members, jedis);
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+         {
+             pool.returnResource( jedis );
+         }
+         return totalKarma;
+    }
+
+    private static long sum(Set<String> members, Jedis jedis) {
+        long sum = 0;
+        for (String member : members) {
+            sum += jedis.zscore(keyOfChurchDonorSortedSet, member);
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+         finally
+         {
+             pool.returnResource( jedis );
+         }
+         return totalKarma;
+    }
+
+    private static long sum(Set<String> members, Jedis jedis) {
+        long sum = 0;
+        for (String member : members) {
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+             totalKarma = sum(members, jedis);
+         }
+         finally
+         {
+             pool.returnResource( jedis );
+         }
+         return totalKarma;
+    }
+
+    private static long sum(Set<String> members, Jedis jedis) {
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: formatting issues of the method

File: vilebot/src/com/oldterns/vilebot/handlers/user/Church.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    private static final Pattern donatePattern = Pattern.compile( "^!donate (-?[0-9]+)\\s*" );
+    private static final Pattern churchTotalPattern = Pattern.compile( "^!churchtotal" );
+    private static final Pattern topDonorsPattern = Pattern.compile( "^!topdonors" );
+    private static final Pattern setTitlesPattern = Pattern.compile( "^!settitle (.+)$" );
+
+    @Handler
+    private void donateToChurch( ReceivePrivmsg event )
+    {
+        Matcher matcher = donatePattern.matcher( event.getText() );
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/handlers/user/Church.java]: remove line

File: vilebot/src/com/oldterns/vilebot/handlers/user/Church.java [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+        if ( matcher.matches() )
+        {
+            String rawDonationAmount = matcher.group( 1 );
+            String donor = BaseNick.toBaseNick( event.getSender() );
+
+            Integer donationAmount = Integer.parseInt( rawDonationAmount );
+            Integer donorKarma = KarmaDB.getNounKarma(donor);
+            donorKarma = donorKarma == null ? 0 : donorKarma;
+
```
Comment (itpun) [inline vilebot/src/com/oldterns/vilebot/handlers/user/Church.java]: remove line

File: vilebot/src/com/oldterns/vilebot/db/ChurchDB.java [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+import java.util.Set;
+
+import redis.clients.jedis.Jedis;
+
+public class ChurchDB
+    extends RedisDB
+{
+	/**
+	 * This class is effectively a duplicate of KarmaDB, used for church related karma
+	 */
```
Comment (amisevsk) [inline vilebot/src/com/oldterns/vilebot/db/ChurchDB.java]: Javadoc on classes goes outside the class definition

**Comment by emmettu** [comment]
Maybe PRs should be run through a style checker to avoid billions of style change requests?

**Comment by yuqi-zhang** [comment]
My bad on that, but it should be fixed now. The non-db has been refractored except changes in Karma.java, which still matches the other code.

**Pull request merged**